### PR TITLE
Fix zero-operand gates and instructions

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

A few checks in different places didn't account for zero-operand gates
being possible.  In most uses, there never would be anything useful
here, but it can occasionally be useful to create a "global-phase"
instruction, especially if this is later going to have controls applied
to it.  This is well-formed within the Qiskit data model; spectator
qubits implicitly undergo the identity, so this corresponds to the
global phase being multiplied by the identity on _all_ qubits.

### Details and comments

This came up in #8236.  This PR doesn't add the requested gate, it just removes the roadblocks from a user-defined definition - such a gate _should_ already be representable in the `Gate.definition` form, but due to the bugs fixed here, would fail in weird ways.

There's probably more places where this edge case will slip through and make a mess that I didn't catch here, but this makes everything I could immediately think of work - transpilation at pre-defined levels, gate control, conversion to `Operator`, etc.

### CODE REVIEW

File: qiskit/circuit/controlledgate.py [hunk 1]
```diff
@@ -174,10 +174,10 @@ def num_ctrl_qubits(self, num_ctrl_qubits):
         Raises:
             CircuitError: num_ctrl_qubits is not an integer in [1, num_qubits - 1].
         """
-        if num_ctrl_qubits == int(num_ctrl_qubits) and 1 <= num_ctrl_qubits < self.num_qubits:
+        if num_ctrl_qubits == int(num_ctrl_qubits) and 1 <= num_ctrl_qubits <= self.num_qubits:
```
Comment (Cryoris) [inline qiskit/circuit/controlledgate.py:177]: We should also update the docs to mention the control qubits must be in `[1, num_qubits]` 😄

**Review by Cryoris — COMMENTED** [review]
It feels a little weird to add a controlled global phase this way but if it's required that's fine by me. So this is to add a global phase on the entire circuit right? If we need a relative phase we could also do something like
```python
phase = QuantumCircuit(1, global_phase=phase)
phase.control(...)
```

File: qiskit/circuit/controlledgate.py (hunk 1 above)
Comment (jakelishman) [reply qiskit/circuit/controlledgate.py:177]: Ah, thanks - I have my editor set to keep docstrings folded by default and missed it!  Fixed in 23ea8950e.

File: test/python/quantum_info/operators/test_operator.py [hunk 2]
```diff
@@ -841,6 +841,13 @@ def test_from_circuit_constructor_empty_layout(self):
         with self.assertRaises(KeyError):
             Operator.from_circuit(circuit, layout=layout)
 
+    def test_compose_scalar(self):
+        """Test that composition works with a scalar-valued operator over no qubits."""
+        base = Operator(np.eye(2, dtype=np.complex128))
+        scalar = Operator(np.array([[-1.0 + 0.0j]]))
```
[the remark this answers is not in the document]
Comment (jakelishman) [reply test/python/quantum_info/operators/test_operator.py:847]: This test is specifically to test composition of two `Operator` instances - I can add a separate one for `ScalarOp` too, but the original behaviour I found and wanted to test was between two direct `Operator` instances.

[the remark this answers is not in the document]
Comment (jakelishman) [reply test/python/quantum_info/operators/test_operator.py:847]: Added a `ScalarOp` test in 361cac062.

**Comment by jakelishman** [comment]
I've fixed the merge conflict and added an extra test for Erick.

**Review by Cryoris — COMMENTED** [review]
Overall LGTM, but now something like 
```python
my_cx = ControlledGate(
    name="cx?", 
    num_qubits=2, 
    params=[],
    num_ctrl_qubits=2, # same as num_qubits!
)
```
is technically a valid input right? Could we add a test to ensure this breaks as expected?

Also in the class docstring is still says that
```
            CircuitError: If ``num_ctrl_qubits`` >= ``num_qubits``.
```
maybe 
```
            CircuitError: If ``num_ctrl_qubits + base_gate.num_qubits`` >= ``num_qubits``.
```
would be more accurate? 🙂

**Comment by jakelishman** [comment]
Oh, good idea on the last part - that's the constraint that actually should be implemented.  We did need the case of `num_qubits == num_ctrl_qubits` to be allowed so that something like `ControlledGate(base_gate=GlobalPhaseGate(pi))` could succeed, but you're right it should still error out for things like `base_gate=XGate()`.  (Of course, the potential `GlobalPhaseGate().control()` would actually just return `PhaseGate()`, but the point stands that the `ControlledGate` constructor _should_ be able to take it.)

**Comment by jakelishman** [comment]
Done in af1abf0cd.

**Review by Cryoris — APPROVED** [review]

**Pull request merged**